File: scala/core/src/main/scala/blockchain/BlockIdTree.scala
package blockchain

import blockchain.models.BlockId
import blockchain.codecs.given
import cats.data.{NonEmptyChain, OptionT}
import cats.effect.{Async, Resource, Sync}
import cats.implicits.*

trait BlockIdTree[F[_]]:
  def parentOf(blockId: BlockId): F[Option[BlockId]]
  def associate(child: BlockId, parent: BlockId): F[Unit]
  def heightOf(t: BlockId): F[Long]
  def findCommonAncestor(
      a: BlockId,
      b: BlockId
  ): F[(NonEmptyChain[BlockId], NonEmptyChain[BlockId])]

object BlockIdTree:
  def make[F[_]: Async](
      read: BlockId => F[Option[(Long, BlockId)]],
      write: (BlockId, (Long, BlockId)) => F[Unit],
      genesisParent: BlockId
  ): Resource[F, BlockIdTree[F]] =
    Resource.pure(new BlockIdTreeImpl[F](read, write, genesisParent))

class BlockIdTreeImpl[F[_]: Async](
    read: BlockId => F[Option[(Long, BlockId)]],
    write: (BlockId, (Long, BlockId)) => F[Unit],
    genesisParent: BlockId
) extends BlockIdTree[F]:

  private def readOrRaise(id: BlockId) =
    OptionT(read(id)).getOrElseF(
      Async[F].raiseError(
        new NoSuchElementException(show"Element not found. id=$id")
      )
    )

  def parentOf(t: BlockId): F[Option[BlockId]] =
    if (t == genesisParent) none[BlockId].pure[F]
    else OptionT(read(t)).map(_._2).value

  def associate(child: BlockId, parent: BlockId): F[Unit] =
    if (parent == genesisParent) write(child, (1, parent))
    else
      readOrRaise(parent).flatMap { case (height, _) =>
        write(child, (height + 1, parent))
      }

  def heightOf(t: BlockId): F[Long] =
    if (t == genesisParent) 0L.pure[F]
    else readOrRaise(t).map(_._1)

  def findCommonAncestor(
      a: BlockId,
      b: BlockId
  ): F[(NonEmptyChain[BlockId], NonEmptyChain[BlockId])] =
    if (a == b) (NonEmptyChain(a), NonEmptyChain(b)).pure[F]
    else
      for {
        (aHeight, bHeight) <- (heightOf(a), heightOf(b)).tupled
        (aAtEqualHeight, bAtEqualHeight) <-
          if (aHeight === bHeight) (NonEmptyChain(a), NonEmptyChain(b)).pure[F]
          else if (aHeight < bHeight)
            traverseBackToHeight(NonEmptyChain(b), bHeight, aHeight).map(
              NonEmptyChain(a) -> _._1
            )
          else
            traverseBackToHeight(NonEmptyChain(a), aHeight, bHeight).map(
              _._1 -> NonEmptyChain(b)
            )
        (chainA, chainB) <- (aAtEqualHeight, bAtEqualHeight).iterateUntilM {
          case (aChain, bChain) =>
            (prependWithParent(aChain), prependWithParent(bChain)).tupled
        } { case (aChain, bChain) => aChain.head == bChain.head }
      } yield (chainA, chainB)

  private def traverseBackToHeight(
      collection: NonEmptyChain[BlockId],
      initialHeight: Long,
      targetHeight: Long
  ): F[(NonEmptyChain[BlockId], Long)] =
    Sync[F].defer(
      (collection, initialHeight)
        .iterateUntilM { case (chain, height) =>
          prependWithParent(chain).map(_ -> (height - 1))
        }(_._2 === targetHeight)
    )

  private def prependWithParent(
      c: NonEmptyChain[BlockId]
  ): F[NonEmptyChain[BlockId]] =
    readOrRaise(c.head).map(_._2).map(c.prepend)

File: dart/protocol/lib/src/blockchain/block_id_tree.dart
import 'package:fixnum/fixnum.dart';
import 'package:giraffe_sdk/sdk.dart';

import 'genesis.dart';

abstract class BlockIdTree {
  Future<BlockId?> parentOf(BlockId blockId);
  Future<void> associate(BlockId child, BlockId parent);
  Future<Int64> heightOf(BlockId blockId);
  Future<(List<BlockId>, List<BlockId>)> findCommonAncestor(
      BlockId a, BlockId b);
}

class BlockIdTreeImpl extends BlockIdTree {
  final Future<(Int64, BlockId)?> Function(BlockId) read;
  final Future<void> Function(BlockId, (Int64, BlockId)) write;

  BlockIdTreeImpl({required this.read, required this.write});

  Future<(Int64, BlockId)> _readOrRaise(BlockId id) {
    return read(id).then((value) {
      if (value == null) {
        throw Exception('Value not found for key=$id');
      }
      return value;
    });
  }

  @override
  Future<void> associate(BlockId child, BlockId parent) async {
    if (parent == Genesis.parentId) {
      await write(child, (Int64(1), parent));
    } else {
      final (parentHeight, _) = await _readOrRaise(parent);
      await write(child, (parentHeight + Int64(1), parent));
    }
  }

  @override
  Future<(List<BlockId>, List<BlockId>)> findCommonAncestor(
      BlockId a, BlockId b) async {
    final chainA = [a];
    final chainB = [b];
    if (a == b) return (chainA, chainB);
    Int64 aHeight = await heightOf(a);
    Int64 bHeight = await heightOf(b);
    while (aHeight > bHeight) {
      a = (await parentOf(a))!;
      chainA.insert(0, a);
      aHeight -= 1;
    }
    while (bHeight > aHeight) {
      b = (await parentOf(b))!;
      chainB.insert(0, b);
      bHeight -= 1;
    }
    while (a != b) {
      a = (await parentOf(a))!;
      chainA.insert(0, a);
      b = (await parentOf(b))!;
      chainB.insert(0, b);
    }
    return (chainA, chainB);
  }

  @override
  Future<Int64> heightOf(BlockId blockId) async {
    if (blockId == Genesis.parentId) {
      return Int64(0);
    } else {
      final (height, _) = await _readOrRaise(blockId);
      return height;
    }
  }

  @override
  Future<BlockId?> parentOf(BlockId blockId) async {
    if (blockId == Genesis.parentId) {
      return null;
    } else {
      final (_, parent) = await _readOrRaise(blockId);
      return parent;
    }
  }
}
